src/ddmd/dcache.d

```diff
@@ -0,0 +1,377 @@
+            snprintf(buf.ptr, buf.length-1, "%s/.dcache-%s", home, keyspace.ptr);
+            printf("%s\n", buf.ptr);
+            fd = open(buf.ptr, O_CREAT | O_RDWR, S_IRWXU);
+            if (fd < 0) perror("Failed to open DCache mapping");
+            if (posix_fallocate(fd, 0, DATA_SIZE + TOC_SIZE + META_SIZE) < 0) perror("Failed to fallocate DCache mapping");
+            
+            mapping = mapRange(META_SIZE + DATA_SIZE + TOC_SIZE, fd);
+            owner = cast(shared(pid_t)*)mapping.ptr;
+            toc = cast(TocEntry[])mapping[META_SIZE .. META_SIZE + TOC_SIZE];
+            data = cast(Data*)(mapping.ptr + META_SIZE + TOC_SIZE);
+            if (toc.length & (toc.length-1)) abort();
+            ourPid = getpid();
+            // Lock before modification of shared data
+            {
+                lock();
+                scope(exit) unlock();
+                // If first entry is all zeros - new mapping
+                if (data.occupied == 0 && data.size == 0)
+                {
+                    data.size = DATA_SIZE;
+                }
+            }
+            // Anything older then tsThreshold is considered ripe for eviction
+            tsCompilation = time(null);
+            tsThreshold = tsCompilation - 24*3600;
+        }
+
+        /// Clean the cache
+        void truncate()
+        {
+            lock();
+            scope(exit) unlock();
+            (cast(ubyte[])mapping)[]  = 0;
+            data.size = DATA_SIZE;
+        }
+
+        static struct TocEntry
+        {
+            uint hash32;
+            uint index;
+            ulong lastUsed; // seconds since 2000-01-01, == 0 - means entry is empty
+        }
+        static assert(TocEntry.sizeof == 16);
+
+        static struct Data
+        {
+            uint occupied; // 0 - free block, 1 - occupied
+            uint size; // size of block, that includes size of this header
+            uint next; // index from the start of data segement
+            char[0] key; // key data, zero-terminated
+            // ubyte[0] - data
+        }
+        static assert(Data.sizeof == 12);
+
+        string get(const(char)[] key)
+        {
+            lock();
+            scope(exit) unlock();
+            uint h = fnv1a(key);
+            size_t offset = h & (toc.length-1);
+            for (;;)
+            {
+                if (toc[offset].lastUsed == 0)
+                {
+                    return null;
+                }
+                if (toc[offset].hash32 == h)
+                {
+                    auto p = cast(Data*)(cast(void*)data + toc[offset].index);
+                    if (p.occupied && 0 == strncmp(p.key.ptr, key.ptr, key.length))
+                    {
+                        auto ptr = p.key.ptr + key.length + 1;
+                        return cast(string)ptr[0 .. p.size - key.length - 1 - Data.sizeof ];
+                    }
+                }
+                offset = (offset + 1) & (toc.length-1);
```

Comment:
Use quadratic probing with triangular numbers, works wonders to avoid (primary) clustering. Table length must be a power of 2, then this series will visit all entries, so you don't even need a stop condition.
https://github.com/dlang/dmd/blob/e2f87ce487985ab12ca4303f4ef1c50cd02f4e70/src/ddmd/root/stringtable.d#L203-L212

also used in druntime

https://github.com/dlang/druntime/blob/fba1d54da42772bbc128f2a72abca286d2010264/src/rt/aaA.d#L115-L122

Reply:
I'm thinking of exploring Robin Hood's hash map. 

Though at the moment I'd keep it dumb maybe even plain rewrite on collision.
With low enough load factor and sane hash function it would work quite well, it's a cache not a strict AA.

Reply:
FWIW Robin Hood hashing is overrated, instead of optimizing access when many collisions occur, it's better to avoid many collisions in the first place.
[Primary Clustering](https://en.wikipedia.org/wiki/Primary_clustering) is a very practical problem, even with MurmurHash. It's easily solved by the trivial quadratic probing scheme above.
Secondary clustering tends to be less of a problem.

https://nbviewer.jupyter.org/urls/dawg.eu/OpenAddressing.ipynb

Also for Robin Hood hashing you need to move entries with could require locking for concurrent access.